aip/general/0193.md

```diff
@@ -75,132 +86,191 @@ following:
-same [concerns](#status-message-warning) regarding client misuse of
-textual error messages apply.
+The [`ErrorInfo`][ErrorInfo] message is the required way to send a
+machine-readable identifier. All error responses **must** include an
+`ErrorInfo` payload in `details`.  Variable information **should** be
+included in `metadata` in `ErrorInfo` and **must** be included if it
+appears within an error message.
 
-#### Dynamic variables
+When introducing an error that represents a failure scenario that did
+not previously occur for the service, the payload **must** include
+`ErrorInfo` and any variables found in dynamic segments of the error
+message **must** be present in `ErrorInfo.metadata` (See [Dynamic
+variables](#dynamic-variables).
+
+**Note:** `ErrorInfo` represents a special case. There **must** be
+exactly one `ErrorInfo`. It is required.
+
+[ErrorInfo Payload][ErrorInfo]
+
+Details of this object are summarized in the following fields, field
+descriptions, and examples:
+
+- `reason string`
+  - A short snake_case description of why the error occurred. Error reasons
+    are unique within a particular domain of errors. The error reason
+    **must** do the following:
+  - Be at most 63 characters and match a regular expression of
+    \`\[A-Z\]\[A-Z0-9\_\]+\[A-Z0-9\]\`, which represents
+    UPPER_SNAKE_CASE.
+  - Be unique and consumable by machine actors for automation.
+  - Be meaningful enough for a human reader to understand what the
+    reason refers to.
+  - Map to one and only one cause, That is, avoid making machine actors
+    reliant on ErrorInfo.
+  - *Example*: CPU_AVAILABILITY
+  - Distill your error message into its simplest form. For example, if
+    your error message is: `The Book, "The Great Gatsby", is unavailable
+    at the Library, "Garfield East". It is expected to be available
+    again on 2199-05-13.` The `reason string` could be one of the
+    following text examples in UPPER_SNAKE_CASE: UNAVAILABLE, NO_STOCK, CHECKED_OUT, AVAILABILITY_ERROR
+  - In contrast, using any of the following reasons breaches the
+    intended formatting and is not recommended: noBooks, THE_BOOK_YOU_WANT_IS_NOT_AVAILABLE, ERROR, librariesAreGreat
+- `domain string`
+  - The logical grouping to which the `reason` belongs. The error domain is
+    typically the registered service name of the tool or product that
+    generated the error. The domain must be a globally unique value.
+  - *Example*: "`pubsub.googleapis.com`"
+- `metadata map`
+  - Additional structured details about this error, which **should** provide
+    important context for customers to identify resolution steps.  Keys
+    **should** match /\[a-zA-Z0-9-\_\]/ and be limited to 64 characters
+    in length. When identifying the current value of an exceeded limit,
+    the units **should** be contained in the key, not the value.
+  - *Example*: `"zone": "us-east1-a", "vmType": "e2-medium", "attachment":
+     "local-ssd=3,nvidia-t4=2"`
```

Comment:
the rendering of these key value pairs is not indicative of a map with the site-generator. Please fix.